src/remoteConnector.ts

```diff
@@ -771,62 +791,91 @@ export default class RemoteConnector extends Disposable {
-				// 		kind,
-				// 		status: 'connected',
-				// 		instanceId: connectionInfo.instanceId,
-				// 		workspaceId: connectionInfo.workspaceId,
-				// 		gitpodHost: connectionInfo.gitpodHost
-				// 	});
-				// } else {
-				// 	this.telemetry.sendTelemetryEvent('vscode_desktop_ssh', {
-				// 		kind,
-				// 		status: 'failed',
-				// 		reason: 'remote-ssh extension: connection failed',
-				// 		instanceId: connectionInfo.instanceId,
-				// 		workspaceId: connectionInfo.workspaceId,
-				// 		gitpodHost: connectionInfo.gitpodHost
-				// 	});
-				// }
-				await this.context.globalState.update(`${RemoteConnector.SSH_DEST_KEY}${sshDestStr}`, undefined);
-			}
+		if (!remoteUri) {
+			return;
+		}
 
-			return connectionSuccessful;
+		const [, sshDestStr] = remoteUri.authority.split('+');
+		const connectionInfo = this.context.globalState.get<SSHConnectionParams & { isFirstConnection: boolean }>(`${RemoteConnector.SSH_DEST_KEY}${sshDestStr}`);
+		if (!connectionInfo) {
+			return;
+		}
+
+		await this.context.globalState.update(`${RemoteConnector.SSH_DEST_KEY}${sshDestStr}`, { ...connectionInfo, isFirstConnection: false });
+
+		const session = await this.getGitpodSession(connectionInfo.gitpodHost);
+		if (session) {
+			this.startHeartBeat(session.accessToken, connectionInfo);
 		}
 
-		return false;
+		// const sshDest = parseSSHDest(sshDestStr);
+		// const connectionSuccessful = await isRemoteExtensionHostRunning();
+		// const usingSSHGateway = typeof sshDest !== 'string';
+		// const kind = usingSSHGateway ? 'gateway' : 'local-app';
+		// if (connectionInfo.isFirstConnection && connectionSuccessful) {
+		// 	this.telemetry.sendTelemetryEvent('vscode_desktop_ssh', {
+		// 		kind,
+		// 		status: 'connected',
+		// 		instanceId: connectionInfo.instanceId,
+		// 		workspaceId: connectionInfo.workspaceId,
+		// 		gitpodHost: connectionInfo.gitpodHost
+		// 	});
+		// } else {
+		// 	this.telemetry.sendTelemetryEvent('vscode_desktop_ssh', {
+		// 		kind,
+		// 		status: 'failed',
+		// 		reason: 'remote-ssh extension: connection failed',
+		// 		instanceId: connectionInfo.instanceId,
+		// 		workspaceId: connectionInfo.workspaceId,
+		// 		gitpodHost: connectionInfo.gitpodHost
+		// 	});
+		// }
+	}
+
+	public override async dispose(): Promise<void> {
+		await this.heartbeatManager?.dispose();
+		super.dispose();
 	}
 }
+
+function isGitpodRemoteWindow(context: vscode.ExtensionContext) {
+	const remoteUri = vscode.workspace.workspaceFile || vscode.workspace.workspaceFolders?.[0].uri;
+	if (vscode.env.remoteName === 'ssh-remote' && context.extension.extensionKind === vscode.ExtensionKind.UI && remoteUri) {
+		const [, sshDestStr] = remoteUri.authority.split('+');
+		const connectionInfo = context.globalState.get<SSHConnectionParams & { isFirstConnection: boolean }>(`${RemoteConnector.SSH_DEST_KEY}${sshDestStr}`);
+
+		return !!connectionInfo;
+	}
+
+	return false;
+}
+
+// async function isRemoteExtensionHostRunning() {
```

Comment:
they are already in git history, it should be safe to remove and you can always use blame to look them up?